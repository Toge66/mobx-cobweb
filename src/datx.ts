/***************************************************
 * Created by nanyuantingfeng on 2020/7/28 09:28. *
 ***************************************************/
export * from '@issues-beta/datx'
