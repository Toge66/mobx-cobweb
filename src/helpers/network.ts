/***************************************************
 * Created by nanyuantingfeng on 2019/11/26 12:22. *
 ***************************************************/
import { IIdentifier, IType, PureCollection, PureModel, View } from '@issues-beta/datx'

import { getCache, saveCache } from './cache'
import { error, getValue, isBrowser } from './utils'

import { IRequestOptions, IRawResponse, IResponseData, $PickOf, ISingleOrMulti, IRequestMethod } from '../interfaces'
import { ResponseView } from '../ResponseView'
import { INetPatchesCollection } from '../interfaces/INetPatchesCollection'

function packResponse<T>(responseData: IResponseData, modelType: IType, collection: PureCollection): IRawResponse<T> {
  const { data = {} as any, ...others } = responseData
  // data : {data : * , meta : *}
  data.type = modelType

  return {
    ...others,
    data,
    collection
  }
}

function getModelEndpointURL(type: IType, collection: PureCollection): string {
  const StaticCollection = collection.constructor as typeof PureCollection
  const QueryModel: any = StaticCollection.types.filter((item) => item.type === type)[0]
  const endpoint = getValue<string>(QueryModel.endpoint)

  if (!endpoint) {
    throw error(`No definition for endpoint was found at Model<${type}>`)
  }

  return endpoint
}

interface IDoFetchOptions {
  collection: INetPatchesCollection<PureCollection> & PureCollection
  options: IRequestOptions
  modelType?: IType
  method: IRequestMethod
  ids?: ISingleOrMulti<IIdentifier>
  views?: View[]
}

async function doFetch<M extends ISingleOrMulti<PureModel>>(doFetchOptions: IDoFetchOptions): Promise<ResponseView<M>> {
  const { options, method = 'GET', collection, views, modelType, ids } = doFetchOptions

  const prepared = collection.adapter.prepare({
    type: modelType,
    endpoint: getModelEndpointURL(modelType, collection),
    ids,
    options,
    method
  })

  const staticCollection = collection && (collection.constructor as { cache?: boolean })
  const collectionCache = staticCollection && staticCollection.cache
  const isCacheSupported = method.toUpperCase() === 'GET'
  const skipCache = doFetchOptions.options && doFetchOptions.options.skipCache

  if (isBrowser && isCacheSupported && collectionCache && !skipCache && prepared.cacheKey) {
    const _response = getCache(prepared.cacheKey, modelType)
    if (_response) {
      // console.info(`cache captured at ${prepared.cacheKey}`)
      return Promise.resolve((_response as unknown) as ResponseView<M>)
    }
  }

  const response1 = await collection.adapter.fetch(prepared.url, prepared.options)
  const response2: IRawResponse<$PickOf<M, object[], object>> = packResponse(response1, modelType, collection)
  const collectionResponse = Object.assign(response2, { collection })
  const response = new ResponseView<M>(collectionResponse, collection, options, undefined, views)

  if (isBrowser && isCacheSupported) {
    saveCache(prepared.cacheKey, modelType, response)
  }

  return response
}

export function query<M extends ISingleOrMulti<PureModel>>(
  modelType: IType,
  options?: IRequestOptions,
  collection?: INetPatchesCollection<PureCollection> & PureCollection,
  views?: View[],
  ids?: ISingleOrMulti<IIdentifier>
): Promise<ResponseView<M>> {
  return doFetch<M>({
    modelType,
    options,
    collection,
    views,
    ids,
    method: 'GET'
  })
}

export function create<T extends PureModel>(
  modelType: IType,
  options?: IRequestOptions,
  collection?: INetPatchesCollection<PureCollection> & PureCollection,
  views?: View[]
): Promise<ResponseView<T>> {
  return doFetch<T>({
    modelType,
    collection,
    options,
    views,
    method: 'POST'
  })
}

export function update<T extends PureModel>(
  modelType: IType,
  options?: IRequestOptions,
  collection?: INetPatchesCollection<PureCollection> & PureCollection,
  views?: View[]
): Promise<ResponseView<T>> {
  return doFetch<T>({
    modelType,
    collection,
    options,
    views,
    method: 'PATCH'
  })
}

export function remove<T extends PureModel>(
  modelType: IType,
  options?: IRequestOptions,
  collection?: INetPatchesCollection<PureCollection> & PureCollection,
  views?: View[]
): Promise<ResponseView<T>> {
  return doFetch<T>({
    modelType,
    collection,
    options,
    views,
    method: 'DELETE'
  })
}
