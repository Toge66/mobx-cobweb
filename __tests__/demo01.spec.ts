import './network'
import { useFixturesByGET } from './nockConfig'

import { collection } from './collection'
import { modelToJSON } from 'datx'
import Staff from './models/Staff'
import Me from './models/Me'

describe('api', () => {
  let scope: any = null

  beforeEach(() => {
    scope = useFixturesByGET()
    collection.removeAll(Staff)
  })

  test('findOne', async () => {
    const sf = await collection.fetch<Staff>(Staff)

    const staff = collection.findOne<Staff>(Staff, 'XRA9koBTaA0000:gongyanyu')
    expect(staff.id).toBe('XRA9koBTaA0000:gongyanyu')
    expect(modelToJSON(staff)).toMatchSnapshot()

    const staffx = collection.findOne<Staff>(Staff.type, 'XRA9koBTaA0000:gongyanyu')
    expect(staffx.id).toBe('XRA9koBTaA0000:gongyanyu')
    expect(staff).toBe(staffx)

    await collection.fetch(Me)
    const me = collection.findAll<Me>(Me)[0]

    expect(modelToJSON(staff)).toMatchSnapshot()
    expect(staff === me.staff).toBeTruthy()

    const staff2 = collection.findOne(Staff, 'XRA9koBTaA0000:gongyanyu')
    expect(staff === staff2).toBeTruthy()

    const staff3 = (sf.data as Staff[]).find(s => s.id === staff.id)

    expect(staff === staff3).toBeTruthy()

    expect(staff.id).toBe('XRA9koBTaA0000:gongyanyu')
    expect(modelToJSON(staff)).toMatchSnapshot()
  })
})
