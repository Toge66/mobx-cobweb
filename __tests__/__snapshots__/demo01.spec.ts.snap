// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api findOne 1`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "userId",
      "id",
      "avatar",
      "email",
      "cellphone",
      "note",
      "external",
      "order",
      "corporation",
      "roles",
    ],
    "id": "XRA9koBTaA0000:gongyanyu",
    "refs": Object {
      "defaultDepartment": Object {
        "model": "organization.Department",
        "type": 0,
      },
      "departments": Object {
        "model": "organization.Department",
        "type": 1,
      },
    },
    "type": "organization.Staff",
  },
  "active": true,
  "avatar": "",
  "cellphone": null,
  "code": "123321",
  "corporation": undefined,
  "corporationId": "XRA9koBTaA0000",
  "createTime": 1561368150495,
  "defaultDepartment": "XRA9koBTaA0000:19031410",
  "departments": Array [
    "XRA9koBTaA0000:19031410",
  ],
  "email": null,
  "external": false,
  "id": "XRA9koBTaA0000:gongyanyu",
  "name": "龚彥宇",
  "nameSpell": "GONGYANYU",
  "note": null,
  "order": Object {
    "XRA9koBTaA0000:19031410": "9043357214046142247",
  },
  "roles": undefined,
  "updateTime": 1574093411607,
  "userId": "tLwu5tpsk7etQAENLoON0giEiE",
  "version": 14,
}
`;

exports[`api findOne 2`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "userId",
      "id",
      "avatar",
      "email",
      "cellphone",
      "note",
      "external",
      "order",
      "corporation",
      "roles",
    ],
    "id": "XRA9koBTaA0000:gongyanyu",
    "refs": Object {
      "defaultDepartment": Object {
        "model": "organization.Department",
        "type": 0,
      },
      "departments": Object {
        "model": "organization.Department",
        "type": 1,
      },
    },
    "type": "organization.Staff",
  },
  "active": true,
  "avatar": "",
  "cellphone": null,
  "code": "123321",
  "corporation": undefined,
  "corporationId": "XRA9koBTaA0000",
  "createTime": 1561368150495,
  "defaultDepartment": "XRA9koBTaA0000:19031410",
  "departments": Array [
    "XRA9koBTaA0000:19031410",
  ],
  "email": null,
  "external": false,
  "id": "XRA9koBTaA0000:gongyanyu",
  "name": "龚彥宇",
  "nameSpell": "GONGYANYU",
  "note": null,
  "order": Object {
    "XRA9koBTaA0000:19031410": "9043357214046142247",
  },
  "roles": undefined,
  "updateTime": 1574093411607,
  "userId": "tLwu5tpsk7etQAENLoON0giEiE",
  "version": 14,
}
`;

exports[`api findOne 3`] = `
Object {
  "__META__": Object {
    "MODEL_PERSISTED_FIELD": true,
    "fields": Array [
      "version",
      "active",
      "createTime",
      "updateTime",
      "name",
      "nameSpell",
      "code",
      "corporationId",
      "userId",
      "id",
      "avatar",
      "email",
      "cellphone",
      "note",
      "external",
      "order",
      "corporation",
      "roles",
    ],
    "id": "XRA9koBTaA0000:gongyanyu",
    "refs": Object {
      "defaultDepartment": Object {
        "model": "organization.Department",
        "type": 0,
      },
      "departments": Object {
        "model": "organization.Department",
        "type": 1,
      },
    },
    "type": "organization.Staff",
  },
  "active": true,
  "avatar": "",
  "cellphone": null,
  "code": "123321",
  "corporation": undefined,
  "corporationId": "XRA9koBTaA0000",
  "createTime": 1561368150495,
  "defaultDepartment": "XRA9koBTaA0000:19031410",
  "departments": Array [
    "XRA9koBTaA0000:19031410",
  ],
  "email": null,
  "external": false,
  "id": "XRA9koBTaA0000:gongyanyu",
  "name": "龚彥宇",
  "nameSpell": "GONGYANYU",
  "note": null,
  "order": Object {
    "XRA9koBTaA0000:19031410": "9043357214046142247",
  },
  "roles": undefined,
  "updateTime": 1574093411607,
  "userId": "tLwu5tpsk7etQAENLoON0giEiE",
  "version": 14,
}
`;
